models/DataObject/ClassDefinition/Data/Relations/AbstractRelations.php

```diff
@@ -121,30 +121,104 @@ public function save(Localizedfield|AbstractData|\Pimcore\Model\DataObject\Objec
         }
 
         $data = $this->getDataFromObjectParam($object, $params);
+        $myCurrentRawRelations = $this->prepareMyCurrentRelations($object, $params);
+        $myNewRawRelations = [];
+        $ignoreClassId = null;
+        $classId = match (true) {
+        	$object instanceof Concrete => $object->getClassId(),
+        	$object instanceof AbstractData => $object->getObject()->getClassId(),
+        	$object instanceof Localizedfield => $object->getObject()->getClassId(),
+        	$object instanceof \Pimcore\Model\DataObject\Objectbrick\Data\AbstractData => $object->getObject()->getClassId(),
+        };
+
+        if (null === $classId) {
+            throw new \Exception('Invalid object type');
+        }
+
         if ($data !== null) {
             $relations = $this->prepareDataForPersistence($data, $object, $params);
 
             if (is_array($relations) && !empty($relations)) {
-                $db = Db::get();
-
                 foreach ($relations as $relation) {
-                    $this->enrichDataRow($object, $params, $classId, $relation);
+                    $this->enrichDataRow($object, $params, $ignoreClassId, $relation);
 
-                    // relation needs to be an array with src_id, dest_id, type, fieldname
-                    try {
-                        $db->insert('object_relations_' . $classId, Db\Helper::quoteDataIdentifiers($db, $relation));
-                    } catch (\Exception $e) {
-                        Logger::error('It seems that the relation ' . $relation['src_id'] . ' => ' . $relation['dest_id']
-                            . ' (fieldname: ' . $this->getName() . ') already exist -> please check immediately!');
-                        Logger::error((string)$e);
+                    if ($object instanceof Concrete) {
+                        $relation['ownername'] = '';  //default in db
+                        $relation['position'] = '0'; //default in db
```

Comment:
The problematic code is not exactly here, but It seem it is not working when the relational field-type is under a Field Collection
![sdvsdf](https://github.com/pimcore/pimcore/assets/6014195/07326f80-90d3-40e5-acd7-c90277ccf3b4)

It would create all and remove all a the same time (`$newRelations` and `$removedRelations` with same row count).

See also the review comment below about ContextPersistenceTrait

Reply:
fixed by casting it to string